File: csharp/Sorts/QuickSort.cs
using System;

namespace Sorts
{
	/// <summary>
	/// Quick sort secret is the Partition. The Partition works very will with nearly
	/// sorted data.  It is an in-place sort.
	/// 
	/// Three indexes involved in the merge. 
	/// </summary>
	public class QuickSort
	{
		public static int Partition<T>(T[] array, int p, int r) where T : IComparable
		{
			int q = p;
			T temp;

			for (int j = p; j < r; j++) 
			{
				if (array [j].CompareTo(array [r]) <= 0) {
					temp = array [j];
					array [j] = array [q];
					array [q++] = temp;
				}
			}

			temp = array [r];
			array [r] = array [q];
			array [q] = temp;

			return q;
		}

		public static void Sort<T>(T[] array, int p, int r) where T : IComparable
		{
			if (p < r) {
				int q = Partition (array, p, r);

				Sort (array, p, q - 1);
				Sort (array, q + 1, r);
			}
		}
	}
}

File: java/Sorts/src/QuickSort.java

/**
 * Quick sort secret is the Partition. It is an in-place sort. The Partition 
 * works very will with nearly sorted data and returns the pivot point.   
 * The pivot point will be sorted correctly in l..r.
 * 
 * Partition, then sort
 * @author bradley
 *
 */
class QuickSort {
	private static <T extends Comparable<T>> int partition(T[] source, int left, int right) {
		T temp;
		int q = left;
		
		for(int i=left; i < right; i++) {
			if(source[i].compareTo(source[right]) < 0) {
				temp = source[i];
				source[i] = source[q];
				source[q++] = temp;
			}
		}
		
		temp = source[right];
		source[right] = source[q];
		source[q] = temp;
		
		return q;
	}
	
	static <T extends Comparable<T>> void sort(T[] source, int left, int right){
		if(left < right) {
			int q = partition(source, left, right);
			//sort about the pivot point
			sort(source, left, q - 1);
			sort(source, q + 1, right);
		}
	}
}
